### PR TITLE
refactor minting method to be more generic and support prefix

### PR DESCRIPTION
**Pull request by rdinicut**
I managed to make the minting method more generic. I had to update truffle in order to use the new ABI encoding for deployment and as a result I had to update all the tests to accommodate the new web3 api.

We still need to change the label names to accommodate the prefix change from https://github.com/centrifuge/go-centrifuge/issues/245

The abi should not change so we can generate the go bindings

### CODE REVIEW

File: test/AnchorRepository.test.js [hunk 1]
```diff
@@ -90,34 +92,24 @@ contract("AnchorRepository", function (accounts) {
...
-            await shouldRevert(anchorRepository.preCommit(anchorId, signingRoot, centrifugeId, null, expirationBlock, callOptions));
-            await shouldRevert(anchorRepository.preCommit(anchorId, signingRoot, centrifugeId, precommitSignature, "", callOptions));
-            await shouldRevert(anchorRepository.preCommit(anchorId, signingRoot, centrifugeId, precommitSignature, null, callOptions));
+            ;
+            await shouldRevert(anchorRepository.preCommit("0x0", signingRoot, centrifugeId, precommitSignature, expirationBlock, callOptions));
+            await shouldRevert(anchorRepository.preCommit(anchorId, "0x0", centrifugeId, precommitSignature, expirationBlock, callOptions));
+            await shouldRevert(anchorRepository.preCommit(anchorId, signingRoot, "0x0", precommitSignature, expirationBlock, callOptions));
+            await shouldRevert(anchorRepository.preCommit(anchorId, signingRoot, centrifugeId, "0x0", expirationBlock, callOptions));
+            await shouldRevert(anchorRepository.preCommit(anchorId, signingRoot, centrifugeId, precommitSignature, "0x0", callOptions));
```
Comment (lucasvo) [inline test/AnchorRepository.test.js]: Were those extra transactions redundant or why did you remove them?

Comment (rdinicut) [reply test/AnchorRepository.test.js]: Yea, it does param type check and it does not allow you to send empty or null.

File: contracts/erc721/PaymentObligation.sol [hunk 2]
```diff
@@ -65,93 +60,41 @@ contract PaymentObligation is UserMintableERC721 {
...
     uint256 _anchorId,
     bytes32 _merkleRoot,
-    string _collaboratorField,
-    string[5] _values,
-    bytes32[5] _salts,
-    bytes32[][5] _proofs
+    string[] _values,
+    bytes32[] _salts,
+    bytes32[][] _proofs,
+    bytes _signature
```
Comment (manuelpolzhofer) [inline contracts/erc721/PaymentObligation.sol]: I think we should refactor our identity to meta transactions together with a proxy functionality. 

Currently, the `AnchorRepository` and `PaymentObligation` have to implement the signature validation.  It should only happen once inside the identity contract.

In the future a `msg.sender` check for `AnchorRepository` and `PaymentObligation` would be enough.

File: contracts/erc721/UserMintableERC721.sol [hunk 3]
```diff
@@ -97,17 +110,36 @@ contract UserMintableERC721 is ERC721Token {
...
+      require(
+        MerkleProofSha256.verifyProof(
+          _proofs[i],
+          _merkleRoot,
+          _hashLeafData(mandatoryFields[i], _values[i], _salts[i])
+        ),
+        mandatoryFields[i]
+      );
+    }
     require(
```
Comment (manuelpolzhofer) [inline contracts/erc721/UserMintableERC721.sol]: I would reorder the `require` statements.
Starting with the lowest cost `require` statement.

In the current setting, a missing anchor registry entry transaction would have to pay the full proof validation before it actually fails.

File: contracts/erc721/PaymentObligation.sol [hunk 4]
```diff
@@ -65,93 +60,41 @@ contract PaymentObligation is UserMintableERC721 {
...
+   * using precise-proofs
+   * @param _salts bytes32[] The salts for the field that is being proved
    * Will be concatenated for proof verification as outlined in
    * precise-proofs library.
-   * @param _proofs bytes32[][5] Documents proofs that are needed
+   * @param _proofs bytes32[][] Documents proofs that are needed
    * for proof verification as outlined in precise-proofs library.
+   * @param _signature signed message used for on chain ACL
    */
   function mint(
```
Comment (manuelpolzhofer) [inline contracts/erc721/PaymentObligation.sol]: Can we remove `PaymentObligation.sol`? In the current implementation, it is just a wrapper for ERC721.
It stores the PODetails but they are stored in ERC721 transactions as well. 

What do you think? I might be cheaper.

Comment (rdinicut) [reply contracts/erc721/PaymentObligation.sol]: It contains the PO logic, the fields, what ever to decide to do with our identities and also getting information about the nft. PODetails are not stored in the ERCBase.

File: contracts/erc721/PaymentObligation.sol (hunk 2 above)
Comment (rdinicut) [reply contracts/erc721/PaymentObligation.sol]: I do not think msg.sender works for a delegate pattern. And in the case of a meta transaction isn't the msg.sender the relayed, proxy contract and you need the signature in order to retrieve the public key?

File: contracts/erc721/PaymentObligation.sol (hunk 4 above)
Comment (manuelpolzhofer) [reply contracts/erc721/PaymentObligation.sol]: Let say I only would use ERC721. 
I could pass the support_fields directly inside the constructor. 

To get the PODetails I could iterate over all transactions for the ERC721 address and looking for correct tokenId in the input.
(Alternative: We could implement it by storing it inside the events)
